src/components/tl/mlx5/alltoall/alltoall.h

```diff
@@ -0,0 +1,194 @@
+#define MAX_BLOCK_SIZE               64 // from limit of Transpose unit capabilities
+#define MAX_OUTSTANDING_OPS          1 //todo change - according to limitations (52 top)
+#define MIN_POLL_WC                  8
+
+typedef uint64_t tl_mlx5_atomic_t;
+typedef uint64_t tl_mlx5_barrier_t;
+
+enum
+{
+    UCC_MLX5_NEED_SEND_MKEY_UPDATE = UCC_BIT(1),
+    UCC_MLX5_NEED_RECV_MKEY_UPDATE = UCC_BIT(2),
+};
+
+typedef struct ucc_tl_mlx5_a2a_ctrl {
+    union {
+        struct {
+            volatile int seq_num;
+            int          mkey_cache_flag;
+        };
+        char tmp[UCC_CACHE_LINE_SIZE];
+    };
+} ucc_tl_mlx5_a2a_ctrl_t;
+
+typedef struct ucc_tl_mlx5_a2a_op {
+    ucc_tl_mlx5_a2a_ctrl_t *ctrl;
+    ucc_tl_mlx5_a2a_ctrl_t *my_ctrl;
+    struct mlx5dv_mkey **   send_mkeys;
+    struct mlx5dv_mkey **   recv_mkeys;
+    int *                   blocks_sent;
+} ucc_tl_mlx5_a2a_op_t;
+
+/* This structure holds resources and data related to the "in-node"
+   part of the algorithm. */
+typedef struct ucc_tl_mlx5_a2a_node {
+    int                  asr_rank;
+    ucc_sbgp_t *         sbgp;
+    void *               storage;
+    ucc_tl_mlx5_a2a_op_t ops[MAX_OUTSTANDING_OPS];
+    struct mlx5dv_mkey * team_recv_mkey;
+    void *               umr_entries_buf;
+    struct ibv_mr *      umr_entries_mr;
+} ucc_tl_mlx5_a2a_node_t;
+
+typedef struct a2a_net_ctrl {
+    struct {
+        void *   addr;
+        uint32_t rkey;
+    } atomic;
+    struct {
+        void *   addr;
+        uint32_t rkey;
+    } barrier;
+} a2a_net_ctrl_t;
+
+typedef struct ucc_tl_mlx5_a2a_net {
+    ucc_sbgp_t *      sbgp;
+    int               net_size;
+    int *             rank_map;
+    ucc_tl_mlx5_qp_t *rc_qps;
+    struct ibv_qp *   dct_qp;
+    struct ibv_srq *  srq;
+    uint32_t *        remote_dctns;
+    struct ibv_ah **  ahs;
+    struct ibv_cq *   cq;
+    struct ibv_cq *   umr_cq;
+    struct ibv_qp *   umr_qp;
+    struct ibv_mr *   ctrl_mr;
+    struct {
+#if ATOMIC_IN_MEMIC
+        struct ibv_dm *counters;
+#else
+        tl_mlx5_atomic_t *counters;
+#endif
+        struct ibv_mr *mr;
+    } atomic;
+    struct {
+        tl_mlx5_barrier_t *flags;
+        struct ibv_mr *    mr;
+    } barrier;
+    int *              blocks_sent; // net_size * MAX_OP_OUTSTANDING
+    a2a_net_ctrl_t *   remote_ctrl;
+    uint32_t *         rkeys;
+    ucc_tl_mlx5_dci_t *dcis;
+} ucc_tl_mlx5_a2a_net_t;
+
+typedef struct ucc_tl_mlx5_a2a_bcast_data {
+    int shmid;
+    int net_size;
+} ucc_tl_mlx5_a2a_bcast_data_t;
+
+enum
+{
+    TL_MLX5_A2A_STATE_SHMID,
+    TL_MLX5_A2A_STATE_EXCHANGE,
+};
+
+typedef struct ucc_tl_mlx5_a2a {
+    struct ibv_pd *              pd;
+    struct ibv_context *         ctx;
+    int                          ib_port;
+    int                          state;
+    int                          transpose;
+    uint64_t                     max_msg_size;
+    ucc_rank_t                   team_size;
```

Comment:
Yes you can, this field is indeed initialized from `UCC_TL_TEAM_SIZE(team)`  in `ucc_tl_mlx5_a2a_init_start`
This is just for convenience, and also for design because we need team_size inside `create_and_populate_recv_team_mkey` whereas this function doesn't take "team" as argument

Reply:
You are right. I removed the field from the struct